In [119]:
import numpy as np
from scipy import stats
import scipy.stats
from scipy.stats import shapiro
import pandas as pd
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
from statsmodels.stats.proportion import proportions_ztest

# 문제 1  
- #### 실습파일 2sample_t_test.csv를 불러온다.
- #### 변수 Mat_A와 Mat_B에 대해 각각 정규성을 검정합니다.

### Mat_A  

- #### 1. 결측치의 유무를 확인한다.
- #### 2. 결측치가 있다면, 결측치를 제거한다,
- #### 3. 'Mat_A'를 기준으로 data를 추출한다.
- #### 4. 추출된 data를 기반으로 정규성 검정을 수행한다.
- #### 5. 정규성 검사 결과를 토대로 t-test 검정을 수행하여, 통계량과 p-value 값을 분석한다.

In [120]:
# 데이터 불러오기
ds_camshaft1 = pd.read_csv('2sample_t_test.csv')

# 결측치 유무 확인
ds_camshaft1.isnull()

Mat_A  Mat_B  Total    Sub
0   False  False  False  False
1   False  False  False  False
2   False  False  False  False
3   False  False  False  False
4   False  False  False  False
5   False  False  False  False
6   False  False  False  False
7   False  False  False  False
8   False  False  False  False
9   False  False  False  False
10   True   True  False  False
11   True   True  False  False
12   True   True  False  False
13   True   True  False  False
14   True   True  False  False
15   True   True  False  False
16   True   True  False  False
17   True   True  False  False
18   True   True  False  False
19   True   True  False  False

In [121]:
# 결측치를 제거한다.
df1 = ds_camshaft1.dropna()
df1

Mat_A  Mat_B  Total  Sub
0   73.4   68.7   73.4    1
1   77.0   71.4   77.0    1
2   73.7   69.8   73.7    1
3   73.3   75.3   73.3    1
4   73.1   71.3   73.1    1
5   71.5   72.7   71.5    1
6   74.5   66.9   74.5    1
7   77.5   70.2   77.5    1
8   76.4   74.4   76.4    1
9   77.7   70.1   77.7    1

In [122]:
# Mat_A에 해당하는 데이터 추출하기
data1 = df1['Mat_A']

In [123]:
# Mat_A 데이터 정규성 검정하기
# 귀무가설 : 정규성을 갖는다.
# 대립가설 : 정규성을 갖지 않는다.
shapiro(data1)

(0.9062638878822327, 0.25630074739456177)

## 결과(1)
    - 'Mat_A'에 대한 정규성 검정 결과, p-value의 값이 유의수준으로 정의된 0.5보다 큰 값을 가지므로 해당 값은 정규성을 가진다는 귀무가설을 채택할 수 있다.

### Mat_B 

- #### 1. 결측치의 유무를 확인한다.
- #### 2. 결측치가 있다면, 결측치를 제거한다,
- #### 3. 'Mat_B'를 기준으로 data를 추출한다.

In [124]:
# 데이터 불러오기
ds_camshaft2 = pd.read_csv('2sample_t_test.csv')

In [125]:
# 결측치 유무 확인
ds_camshaft2.isnull()

Mat_A  Mat_B  Total    Sub
0   False  False  False  False
1   False  False  False  False
2   False  False  False  False
3   False  False  False  False
4   False  False  False  False
5   False  False  False  False
6   False  False  False  False
7   False  False  False  False
8   False  False  False  False
9   False  False  False  False
10   True   True  False  False
11   True   True  False  False
12   True   True  False  False
13   True   True  False  False
14   True   True  False  False
15   True   True  False  False
16   True   True  False  False
17   True   True  False  False
18   True   True  False  False
19   True   True  False  False

In [126]:
# 결측치를 제거한다.
df2 = ds_camshaft2.dropna()
df2

Mat_A  Mat_B  Total  Sub
0   73.4   68.7   73.4    1
1   77.0   71.4   77.0    1
2   73.7   69.8   73.7    1
3   73.3   75.3   73.3    1
4   73.1   71.3   73.1    1
5   71.5   72.7   71.5    1
6   74.5   66.9   74.5    1
7   77.5   70.2   77.5    1
8   76.4   74.4   76.4    1
9   77.7   70.1   77.7    1

In [127]:
# Mat_B에 해당하는 데이터 추출하기
data2 = df2['Mat_B']

# Mat_B 데이터 정규성 검정하기
# 귀무가설 : 정규성을 갖는다.
# 대립가설 : 정규성을 갖지 않는다.
shapiro(data2)

(0.9729756116867065, 0.9169799089431763)

## 결과 (2)
    - 'Mat_B'에 대한 정규성 검정 결과, p-value의 값이 유의수준으로 정의된 0.5보다 큰 값을 가지므로 해당 값은 정규성을 가진다는 귀무가설을 채택할 수 있다.

___
# 문제 2  
- #### 새로운 품종의 감자 20Box를 랜덤추출하여 무게를 측정할 결과가 다음과 같다.
- #### 단, 모 표준편차는 3.8로 알려져 있음, 신품종 감자의 평균 무게에 대한 95% 신뢰구간을 구하시오
- #### 99 page

In [128]:
df = pd.DataFrame({"sample":[18, 18, 20, 21, 20, 23, 19, 18, 17, 21, 22, 20, 20, 21, 20, 19, 19, 18, 17, 19]})
df.head()

sample
0      18
1      18
2      20
3      21
4      20

In [129]:
lower, upper = stats.norm.interval(0.95, loc=np.mean(df), scale=0.38/np.sqrt(20))
print("신뢰구간 : ({0}, {1})".format(lower.round(2), upper.round(2)))

신뢰구간 : ([19.33], [19.67])


## 결과
    - 신품종 감자의 평균 무게에 대한 신뢰구간은 [19.33, 19.67] 이다

___
# 문제 3  
- #### 제품개발자는 새로 개발된 강종의 평균 인장강도를 70km/mm^2으로 기대하고 있음
- #### 이를 검증하기 위해 시험 생산되는 강종 중 임의로 20개의 코일을 대상으로 랜덤하게 상기 자료부터
- #### 표본 평균은 71.50 kg/mm^2, 표본표준편차는 1.9333 kg/mm^2 임을 알았음
- #### 이 모평균값의 95% 신뢰구간을 추정하시오.
- #### 102 page

In [130]:
df = pd.DataFrame({'sample':[73, 71, 74, 69, 70, 73, 70, 68, 75, 72, 70, 72, 73, 70, 70, 72, 71, 70, 75, 72]})

lower, upper = stats.t.interval(0.95, len(df)-1, loc=np.mean(df), scale=scipy.stats.sem(df))
print("신뢰구간 : ({0},{1})".format(lower.round(2), upper.round(2)))

신뢰구간 : ([70.6],[72.4])


# 결과
- 강종의 평균 인장강도에 대한 신뢰구간은 [70.6, 72.4] 이다

___
# 문제 4  
- #### 고객 서비스센터의 고객만족도 평균은 78.0 이다.
- #### 개선활동을 완료한 후 다음과 같은 10개의 고객만족도 데이터를 얻었다.
- #### 개선활동이 만족도를 변화시켰는가? (유의수준 a = 0.05)
- #### 131 page


In [131]:
# 데이터 입력
df = pd.DataFrame({'sample':[85.0, 79.0, 79.1, 79.9, 81.6, 78.6, 85.4, 83.4, 78.1, 79.2]})

# 정규성 검정
shapiro(df)

(0.8473217487335205, 0.0539751797914505)

#### => 귀무가설 : 이미 정의되어 있는 가설, 여기서는 "평균은 78.0" 이다.
#### => 대립가설 : 증명하여 밝히고자 하는 가설, 여기서는 "평균은 78.0이 아니다" 이다.
#### => p-value의 값이 유의수준인 0.05보다 작으므로, 해당 데이터는 정규성을 갖는다는 귀무가설을 기각한다고 말할 수 있다.
#### => 따라서, 평균은 78.0이 아니라고 할 수 있다.

In [132]:
# t-test를 통해 평균 검정
t_result = stats.ttest_1samp(df, 78.0)

# t(통계량)와 p(p-value)값을 변수에 각각 정함
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("Sample t-test")
print("t 통계량 : {0}".format(t))
print("p-value : {0}".format(p))

Sample t-test
t 통계량 : [3.379]
p-value : [0.008]


# 결과
- p-value의 값이 유의수준인 0.05보다 작으므로, 해당 데이터는 정규성을 갖는다는 귀무가설을 기각한다고 말할 수 있다.
- 그렇기에 평균은 78.0이 아니라고 말할 수 있으며, 통계량이 양수이므로 평균이 78.0보다 더 크다고 말할 수 있다.
- 따라서, 개선활동으로 인해서 만족도가 상승했다고 말할 수 있다.

___
# 문제 5  
- #### 제품을 A공급자와 B공급자로 부터 제품을 받아 두께를 측정해서 다음의 데이터를 얻었다.
- #### 공급자 간에 모평균에 차이가 있다고 할 수 있는가? (유의수준 a = 0.05)
- #### 139 page

In [137]:
df1 = pd.DataFrame({'Sample':[1.883,1.715,1.799,1.768,1.711,1.832,1.427,1.344]})
df2 = pd.DataFrame({'Sample':[1.435,1.572,1.486,1.511,1.457,1.548,1.404,1.883]})

shapiro(df1)

(0.8425756096839905, 0.08004552870988846)

- #### p-value의 값이 유의수준보다 크므로, 정규성을 갖는다고 할 수 있다.

In [141]:
shapiro(df2)

(0.7734811902046204, 0.014793740585446358)

- #### p-value의 값이 유의수준보다 작으므로, 정규성을 갖는다고 할 수 없다.

In [142]:
stats.levene(df1['Sample'], df2['Sample'])

LeveneResult(statistic=0.4408416728536036, pvalue=0.5175025061092301)

- #### p-value의 값이 유의수준보다 크므로, 등분산성을 갖는다고 할 수 있다.

In [143]:
t_result = stats.ttest_ind(df1, df2)
t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("Sample t-test")
print("t 통계량 : {0}".format(t))
print("p-value : {0}".format(p))

Sample t-test
t 통계량 : [1.7]
p-value : [0.111]


# 결과
- p-value의 값이 유의수준인 0.05보다 크므로, 귀무가설인 "공급자 간 모평균에 차이가 있다"가 입증된다고 말할 수 있다.
- 또한, 통계량이 양수의 값을 가지므로 A의 모평균이 더 크다고 할 수 있으며, 이는 곧 A공급자로 부터 받은 두께의 평균이 더 크다고 할 수 있다.

___
# 문제 6  
- #### 회사원 10명에 대해 영어점수가 학원가기 전의 점수와 학원 수강 후에 영어점수가 짝을 이루고있다.
- #### 회사원들의 성적은 학원가기 전후에 대해 차이가 있는지 검정하라
- #### 142 page

In [147]:
df1 = pd.DataFrame({'Before':[720, 589, 780, 648, 720, 589, 780, 648, 780, 648]})
df2 = pd.DataFrame({'After':[710, 580, 787, 712, 750, 600, 782, 670, 790, 680]})

In [148]:
shapiro(df1)

(0.871309757232666, 0.10353939980268478)

- #### p-value의 값이 유의수준보다 크므로, 정규성을 갖는다고 할 수 있다.

In [149]:
shapiro(df2)

(0.9132594466209412, 0.304139107465744)

- #### p-value의 값이 유의수준보다 크므로, 정규성을 갖는다고 할 수 있다.

In [150]:
stats.levene(df1['Before'], df2['After'])

LeveneResult(statistic=0.20658136698863727, pvalue=0.6548936313011526)

- #### p-value의 값이 유의수준보다 크므로, 등분산성을 갖는다고 할 수 있다.

In [151]:
t_result = stats.ttest_rel(df1, df2)

t, p = t_result.statistic.round(3), t_result.pvalue.round(3)

print("Paired t-test")
print("t 통계량 : {0}".format(t))
print("p-value : {0}".format(p))

Paired t-test
t 통계량 : [-2.266]
p-value : [0.05]


# 결과
- #### p-value의 값이 유의수준과 같기에 귀무가설이 성립된다고 할 수 있다.
- #### 해당 문제에서의 귀무가설 : 공급자 간 모평균에 차이가 존재한다.
- #### 또한, 통계량을 살펴본 결과 음수의 값을 가지므로, After의 평균이 더 크다는 것을 알 수 있고,
- #### 이는 B 공급자에게서 받은 제품의 두께가 더 두껍다는 것을 의미한다.